Serve Word Vectors

papajohn opened the pull request.

Commit:
initial word2vec support

Commit:
adds similar matches to protocol

Commit:
adds concordance ranking and switches to oscii-lab/Word2VecJava

Commit:
catch NaN numbers in JSON response

Commit:
optimization (move context calculation out of loop)

Commit:
fix failed build

Commit:
support of multiple word2vec models through Word2VecManager class

Commit:
bugfix String array from set

Commit:
skip word2vec concordance ranking on empty context + move dbg output

Commit:
tweak thresholds

papajohn committed:
adopt env convention

papajohn commented on settings.gradle:
```diff
@@ -10,3 +10,6 @@ project(":CoreNLP").projectDir = new File("$System.env.CORENLP_HOME")
 
+// Word2VecJava, note this has to be in the same parent directory as lex
+include "Word2VecJava"
+project(":Word2VecJava").projectDir = new File(System.getProperty("user.dir") + "/../Word2VecJava")
```
I switched this to the environment variable convention. Parallel directories are rejected by my IDE, and also I think we should try to be consistent.

sahasan replied:
no problem, I was between relative path and environment variable, and decided that I didn't want to force people to set up another env var; guess it's better this way

papajohn commented on src/main/java/org/oscii/concordance/Word2VecManager.java:
```diff
@@ -0,0 +1,262 @@
+   * @param lang
+   * @param query
+   */
+  public List<Double> getRawVector(String lang, String query) throws UnknownWordException {
```
If this is somewhere in your inner loop (I don't think it is), then switch to double[] so that you don't get a bunch of Double objects.

sahasan replied:
no inner loop served from this, it's just for the service request `wordvec=true` which serves the raw vector for a query

papajohn commented on src/main/java/org/oscii/concordance/Word2VecManager.java:
```diff
@@ -0,0 +1,262 @@
+    }
+  }
+
+	/**
```
Indentation is wonky from here to the end.

sahasan replied:
ah tabs! from copy-paste, fixed

papajohn commented on src/main/java/org/oscii/lex/Order.java:
```diff
@@ -47,4 +48,11 @@ public int compare(PhrasalRule x, PhrasalRule y) {
         }
     };
+
+    public static final Comparator<? super SentenceExample> bySimilarity = new Comparator<SentenceExample>() {
```
`... bySimilarity = (o1, o2) -> Double.compare(o2.similarity, o1.similarity)`

Commit:
create forSearch() instances only once

the reason for the slowdown is forSearch() calculating a normalized
model every time it was called; now, this normalization happens only
once

Commit:
adds limit flag to defer limiting of concordance lists for word2vec

also increases filtering thresholds since word2vec is now fast due to
previous change

Commit:
adds more info logging

Commit:
strip punctuation from context (since it might be passed in raw)

Commit:
better warn message on unsuccessful word2vec

papajohn commented on README.md:
````diff
@@ -2,6 +2,16 @@
+You have to clone the following repositories as a pre-requisite:
+
+```
+ $ git clone git@github.com:stanfordnlp/phrasal.git
````
phrasal and stanfordnlp are not dependencies, as far as I know.

sahasan replied:
I thought they were since they are part of `settings.gradle`

papajohn commented on src/main/java/org/oscii/Main.java:
```diff
@@ -78,8 +79,14 @@ public static void main(String[] args) throws Exception {
-        
-        final LexiconProtocol protocol = new LexiconProtocol(lexicon, corpus, ranker);
+
+        Word2VecManager word2vecMan = null;
```
Can we call it `embeddings`?

sahasan replied:
sure, do you want it to be an `EmbeddingsManager` as well? I wasn't very creative to be honest

papajohn replied:
Word2VecManager seems fine.

papajohn commented on src/main/java/org/oscii/Main.java:
```diff
@@ -120,7 +127,11 @@ private static OptionSet parse(String[] args) throws IOException {
+
+        // Word2Vec
+        parser.accepts("word2vec", "path to binary Word2Vec model file").withRequiredArg().ofType(File.class);
+        parser.accepts("srclang", "language of binary Word2Vec model (default: en)").withRequiredArg().defaultsTo("en");
```
I thought we could now have multiple languages loaded at once.

sahasan replied:
yes, that's in LexService, let me push that as well here

papajohn commented on src/main/java/org/oscii/api/LexiconProtocol.java:
```diff
@@ -20,13 +28,15 @@
     private final Lexicon lexicon;
     private final AlignedCorpus corpus;
     private final Ranker ranker;
+    private final Word2VecManager word2vecManager;
```
I'd prefer `embeddings` as a name, if possible.

papajohn commented on src/main/java/org/oscii/api/LexiconProtocol.java:
```diff
@@ -42,6 +52,9 @@ public Response respond(Request request) {
         if (request.example) addExamples(request, response);
         if (request.extend) addExtensions(request, response);
         if (request.synonym) addSynonyms(request, response);
+        if (request.wordvec) addWordVector(request, response);
```
And I would have called it `request.embedding` and `addEmbedding`.

papajohn replied:
Also, there are embeddings of whole phrases (not just words) in the Google News corpus, right? Do we support looking those up?

sahasan replied:
you're right! lookup is supported if you formulate the query accordingly (e.g. `San_Francisco`), but ranking of concordance entries does not take those into account; this would require extraction of bigrams/trigrams in the context and the sentence examples (not sure if the trigrams are part of the model, need to check) and score them instead of the unigram matches.

papajohn commented on src/main/java/org/oscii/api/LexiconProtocol.java:
```diff
@@ -80,15 +93,26 @@ private void addDefinitions(Request request, Response response) {
-        // TODO Rank examples (e.g., based on request.context)
-        //  - https://github.com/lilt/core/issues/97
+        long startTime = System.nanoTime(); // - startTime) / 1e9;
+        boolean bHasWord2Vec = word2vecManager.hasModels();
```
I think if no word2vec file is specified on the command line, then `word2vecManager` is `null` at this point, so you'll get an exception here. Didn't try it, though.

papajohn commented on src/main/java/org/oscii/api/LexiconProtocol.java:
```diff
@@ -139,6 +163,71 @@ private void addSynonyms(Request request, Response response) {
+     * Adds the raw word vector for a query to the response.
+     *
+     * Example:
+     * http://localhost:8090/translate/lexicon?query=explain&wordvec=true
```
If you change request `wordvec`->`embedding`, then this example needs to change too.

papajohn commented on src/main/java/org/oscii/concordance/Word2VecManager.java:
```diff
@@ -0,0 +1,268 @@
+  }
+
+  /**
+   * Ranks a list of cocordances by calculating the similarity of each
```
concordance entries

papajohn commented on src/main/java/org/oscii/concordance/Word2VecManager.java:
```diff
@@ -0,0 +1,268 @@
+ * Class that manages a collection of language-specific word2vec models.
+ */
+public class Word2VecManager {
+  private static final int MIN_SEG_LEN = 25; // minimum segment length to activate reduction
```
All other files in `lex` indent with 4 spaces.

papajohn commented:
Aside from comments above, looks good to me! Merge when ready.

papajohn commented:
@ahaas: will this interface work with your synonym visualizer?

ahaas commented:
@papajohn Is there a way to retrieve vectors in bulk? I.e. can I send a request with an array of words, and receive a response with an array of vectors?

papajohn commented:
Currently I think no, but let's add that. It can be a separate pull
request, perhaps.

On Thu, Dec 10, 2015 at 3:02 PM, Andre Haas notifications@github.com
wrote:

> Is there a way to retrieve vectors in bulk? I.e. can I send a request with
> an array of words, and receive a response with an array of vectors?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/oscii-lab/lex/pull/13#issuecomment-163780795.

ahaas commented:
Okay. My synonym visualizer often needs to lookup dozens or hundreds of words at once and sending that number of requests would be inefficient.

Commit:
addresses code review comments (renaming, indentation, small fixes)

sahasan commented:
I suggest the following to be addressed in a separate pull request:
- support for looking up multiple embeddings in one request
- adding bigram/trigram extraction and scoring to concordance ranking
